packages/utils/crypto/src/encryption.ts

```diff
@@ -1,19 +1,218 @@
 export * from "./errors.js";
+
 import {
 	AbstractType,
 	deserialize,
 	field,
 	serialize,
 	variant,
-	vec
+	vec,
+	fixedArray
 } from "@dao-xyz/borsh";
 import { equals } from "@peerbit/uint8arrays";
 import { AccessError } from "./errors.js";
 import sodium from "libsodium-wrappers";
 import { X25519Keypair, X25519PublicKey, X25519SecretKey } from "./x25519.js";
-import { Ed25519Keypair, Ed25519PublicKey } from "./ed25519.js";
+import { Ed25519PublicKey } from "./ed25519.js";
 import { randomBytes } from "./random.js";
-import { Keychain } from "./keychain.js";
+import { sha256 } from "./hash.js";
+
+export type PublicKeyEncryptionParameters = {
+	type?: "publicKey";
+	receiverPublicKeys: (X25519PublicKey | Ed25519PublicKey)[];
+};
+
+export type SymmetricKeyEncryptionParameters = {
+	type?: "hash";
+};
+/* 
+export type NoExchange = {
+	type: 'none'
+};
+ */
+
+export type KeyExchangeOptions =
+	| PublicKeyEncryptionParameters
+	| SymmetricKeyEncryptionParameters;
+
+type EncryptReturnValue<
+	T,
+	Parameters extends KeyExchangeOptions
+> = EncryptedThing<T, EnvelopeFromParameter<Parameters>>;
+
+type CipherWithEnvelope<E = PublicKeyEnvelope | HashedKeyEnvelope> = {
+	cipher: Uint8Array;
+	nonce: Uint8Array;
+	envelope: E;
+};
+
+type SymmetricKeys = Uint8Array;
+type PublicKeyEncryptionKeys = X25519Keypair;
```

Comment:
This should only take a public key IMO. That way we can also encrypt with public keys where we don't own the secret key